src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1220,123 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool isTypeDeclaration)
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.RefKeyword);
+
+            var nextToken = PeekToken(1);
+
+            isRefFollowedByStructOrPartialStruct = nextToken.Kind == SyntaxKind.StructKeyword ||
+                (nextToken.ContextualKind == SyntaxKind.PartialKeyword && this.PeekToken(2).Kind == SyntaxKind.StructKeyword);
+
+            if (isRefFollowedByStructOrPartialStruct)
+            {
+                isTypeDeclaration = true;
+                return true;
+            }
+
+            isTypeDeclaration = IsTypeDeclaration(out nextToken);
+            return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
+        }
+
+        private static bool CanFollowTypeName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.DotToken:                   // partial.     qualified
+                case SyntaxKind.ColonColonToken:            // partial::    alias
+                case SyntaxKind.LessThanToken:              // partial<     generic type
+                case SyntaxKind.OpenBracketToken:           // partial[     array
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool CanFollowMemberName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.LessThanToken:              // partial<     generic method
+                case SyntaxKind.OpenBraceToken:             // partial {    property
+                case SyntaxKind.EqualsGreaterThanToken:     // partial =>   property
+                case SyntaxKind.OpenParenToken:             // partial(     method
+                    return true;
+            }
+
+            return false;
+        }
+
+        private enum PartialModFlags
+        {
+            PartialType,
+            PartialMember,
+            TreatAsModifier,
+            NotModifier
+        }
+
+        private PartialModFlags ShouldPartialBeTreatedAsModifier()
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
+
+            int peekIndex = 1;
+            SyntaxToken currentToken;
+
+            while (GetModifier(currentToken = PeekToken(peekIndex)) != DeclarationModifiers.None)
+            {
+                if (SyntaxFacts.IsContextualKeyword(currentToken.ContextualKind))
```

Comment:
this seems super broad.  You're accepting *any* contextual keyword? so.. like if the user uses "from"?

Reply:
only contextual modifiers. `while` condition guard against others.

Reply:
i suppose it has to be something that is actually a modifier... still... this seems broad and confusing.  I'm guessing it's for async/partial?  If so, perhaps just check for that, or comment as such.

Reply:
note: I'm not sure if this even worth it, ShouldAsyncBeTreatedAsModifier just expects a single additional partial mod. and still it's incomplete, for example, we could not know the second partial in `partial partial<T>` is a type name or a member name unless for look much further.

Reply:
So my take on this is: 

I think the existing parser code is trying to be way to 'slim' and goes through incredibly complex hoops to try to figure out what's going on.  A *much* simpler approach, for me, would be to

1. mark the starting point of the potential modifiers.
2. eagerly consume *anything* that could be the modifiers of a type/member, and keep track of if you saw contextual modifiers, and what position they were. 
3. If you didn't see partial/async (or any other contextual modifiers we add in the future), great! Just return the modifiers you parsed out.  Do not reset, just release your rewind point.
4. if you did have contextual modifiers, now actually examine them to ensure they're ok.  this is hte point where we can do complex checks between them and hte other modifiers, and/or the tokens that follow.
5. If you realize that one of these contextual modifiers was not actually a modifier, store that index.  Rewind.  Consume modifiers up to that part, then actually go do the rest of the parsing.

This just seems much simpler and saner for me.  it feels like right now we're trying to eat a token at a time, trying to keep track of everything and doing complex incremental moving forward and peeking.  I'd just rather get the modifiers, and hten see if partial/async are used properly based on if there are other modifiers that follow and/or you can properly see the start of a member/type afterwards.

Reply:
I feel like we're extremely averse to anything that may rewind, when it's actually the case that rewinding should almost never happen.  you would literally only rewind in the case that partial/async were used as an actual type/member-name.  And i think we can state pretty confidently that that's like the 0.0001% case.  This approach would be simple and fast for the mainline case, and give us a good point at which we could easily check for the rare cases with all the information available in the right context.

Reply:
>This just seems much simpler and saner for me.

The approach you're described is just what I had in mind. thing is that things can get too complicated when you want to be precise. My question is that if all this would worth it -- those costly situations only happens with partial methods and end up with error for 99% cases.

Reply:
IMO, almost anything would be better than what we have :)    I'd be happy to try it out myself, but i won't have any time until the weekend.  

> My question is that if all this would worth it --

IMO... yes.  Right now modifier parsing is suhc a PITA, and has had numerous bugs in it over the years.  Indeed, it seems like every time it is touched, some thing get fixed and some things regress.  i think it should be burned to the ground and just made simple and clear.